lib/ai/model/agent.go

```diff
@@ -0,0 +1,319 @@
+		}
+
+		iterations++
+	}
+}
+
+type stepOutput struct {
+	finish      *AgentFinish
+	action      *AgentAction
+	observation string
+	tokensUsed  int
+}
+
+func (a *Agent) takeNextStep(ctx context.Context, state *executionState) (stepOutput, error) {
+	log.Debug("agent entering takeNextStep")
+	defer log.Debug("agent exiting takeNextStep")
+
+	action, finish, err := a.plan(ctx, state)
+	if err, ok := trace.Unwrap(err).(*invalidOutputError); ok {
+		log.Debugf("agent encountered an invalid output error: %v, attempting to recover", err)
+		action := &AgentAction{
+			action: actionException,
+			input:  observationPrefix + "Invalid or incomplete response",
+			log:    thoughtPrefix + err.Error(),
+		}
+
+		// The exception tool is currently a bit special, the observation is always equal to the input.
+		// We can expand on this in the future to make it handle errors better.
+		log.Debugf("agent decided on action %v and received observation %v", action.action, action.input)
+		return stepOutput{action: action, observation: action.input}, nil
+	}
+	if err != nil {
+		log.Debugf("agent encountered an error: %v", err)
+		return stepOutput{}, trace.Wrap(err)
+	}
+
+	// If finish is set, the agent is done and did not call upon any tool.
+	if finish != nil {
+		log.Debug("agent picked finish, returning")
+		return stepOutput{finish: finish}, nil
+	}
+
+	var tool Tool
+	for _, candidate := range a.tools {
+		if candidate.Name() == action.action {
+			tool = candidate
+			break
+		}
+	}
+
+	if tool == nil {
+		log.Debugf("agent picked an unknown tool %v", action.action)
+		action := &AgentAction{
+			action: actionException,
+			input:  observationPrefix + "Unknown tool",
+			log:    thoughtPrefix + "No tool with name " + action.action + " exists.",
+		}
+
+		return stepOutput{action: action, observation: action.input}, nil
+	}
+
+	if tool, ok := tool.(*commandExecutionTool); ok {
+		input, err := tool.parseInput(action.input)
+		if err != nil {
+			action := &AgentAction{
+				action: actionException,
+				input:  observationPrefix + "Invalid or incomplete response",
+				log:    thoughtPrefix + err.Error(),
+			}
+
+			return stepOutput{action: action, observation: action.input}, nil
+		}
+
+		completion := &CompletionCommand{
+			Command: input.Command,
+			Nodes:   input.Nodes,
+			Labels:  input.Labels,
+		}
+
+		log.Debugf("agent decided on command execution, let's translate to an AgentFinish")
+		return stepOutput{finish: &AgentFinish{output: completion}}, nil
+	}
+
+	return stepOutput{}, trace.NotImplemented("")
+}
+
+func (a *Agent) plan(ctx context.Context, state *executionState) (*AgentAction, *AgentFinish, error) {
+	scratchpad := a.constructScratchpad(state.intermediateSteps, state.observations)
+	prompt := a.createPrompt(state.chatHistory, scratchpad, state.humanMessage)
+	resp, err := state.llm.CreateChatCompletion(
+		ctx,
+		openai.ChatCompletionRequest{
+			Model:    openai.GPT4,
```

Comment:
Can we extract the model name to a constant to avoid repeating it in every initialization?

Reply:
What do you mean? I'm not sure I understand the question here, really? `openai.GPT4` is already a constant and we have one callsite to `CreateChatCompletion`. What are we looking to accomplish with what change?

Reply:
I was thinking about `const usedModel = openai.GPT4`, so you can easily switch it in all places, but more I think about it less sense that make 😅